proof/refine/ARM/KHeap_R.thy

```diff
@@ -74,65 +74,187 @@ abbreviation (input)
 where
   "set_obj' ptr obj s \<equiv> set_ko' ptr (injectKO obj) s"
 
+(* FIXME: move *)
+(** reader monad **)
+
+(* if state s satisfies P, and if f s returns a value, the value satisfies Q *)
+abbreviation
+  ovalid :: "('s \<Rightarrow> bool) \<Rightarrow> ('s,'a) lookup \<Rightarrow> ('a \<Rightarrow> 's \<Rightarrow>  bool) \<Rightarrow> bool"
+  ("o\<lbrace>_\<rbrace>/ _ /\<lbrace>_\<rbrace>")
+where
+  "o\<lbrace>P\<rbrace> f \<lbrace>Q\<rbrace> \<equiv> \<forall>s. P s \<longrightarrow> (\<exists>y. f s = Some y) \<longrightarrow>  Q (the (f s)) s"
+
+abbreviation (input)
+  oinvariant :: "('s,'a) lookup \<Rightarrow> ('s \<Rightarrow> bool) \<Rightarrow> bool" ("_ o\<lbrace>_\<rbrace>" [59,0] 60)
+where
+  "oinvariant f P \<equiv> o\<lbrace>P\<rbrace> f \<lbrace>\<lambda>_. P\<rbrace>"
+
+lemma asks_wp[rule_format, rotated, simp]:
+  "o\<lbrace>\<lambda>s. P (f s) s\<rbrace> asks f \<lbrace>P\<rbrace>"
+  by(simp add:  split_def asks_def oreturn_def obind_def)
+
+(* no fail for reader monad *)
+definition
```

Comment:
Please move to `lib/Monad_WP/OptionMonadWP` or `lib/Monad_WP/OptionMonad` before merge.

Reply:
Actually, `no_ofail` already exists as well in `OptionMonadWP`, so we should use that and prove the additional rules there.

Reply:
oh, great, will do that. I probably should have studied those files a bit more before getting started 😄

Reply:
It seems that definitions in `OptionMonadWP` are not accessible from Refine files, while those in `OptionMonad` are. I am not sure why it is that way.

Reply:
Ah, that explains why they didn't show up.

That's entirely possibly, because AInvs on ARM doesn't use the option monad much or at all, so only the most basic bits would have been imported. We should include OptionMonadWP somewhere in the base AInvs inclusions, I think, so it is available by default. This might theoretically break something, so needs to be checked, but it should be fine if we didn't use the option monad.

Reply:
Ok, I’ll chase it up.

Reply:
So, ASpec imports `OptionMonad` only, via `Machine_A`: `OptionMonad` -> `OptionMonadND` -> `ExecSpec.MachineTypes` -> `Machine_A`

`OptionMonadND` imports `OptionMonad`, and `OptionMonadWP` imports `OptionMonadND`.

We could change `MachineTypes` to import `OptionMonadWP` instead of `OptionMonadND`.

But from what you are suggesting the other comment, we could import `OptionMonadWP`  in `HaskellLib`, which seems like a better option.

Reply:
Yes, I think that's good. The monad constants should be defined in `OptionMonad`, which is included in `ASpec`, and validity, lemmas, etc, should be in `OptionMonadWP/ND`, which makes sense in `Haskell_Lib`.

Reply:
> We could change `MachineTypes` to import `OptionMonadWP` instead of `OptionMonadND`.
> 
> But from what you are suggesting the other comment, we could import `OptionMonadWP` in `HaskellLib`, which seems like a better option.

This gives an issue coming from the syntax overloading of `bind`, which seems obscure.

Reply:
I can have a look at that if you like. If the general `bind` is in scope, usually you can disambiguate with `NonDetMonad.bind` or things like that, or by annotating the type.

Reply:
That did work. Thanks, Gerwin!

Reply:
Right, now that we have imported `OptionMonadWP` in Refine, I can see that there is `ogets` defined in `OptionMonadND` which is equivalent to `asks`. Shall we replace `asks` with `ogets`?

Reply:
Oh, well spotted, I overlooked that one. Should have been defined in `OptionMonad` of course. I blame DavidG :-)

I guess we have to keep the old one, because it is exported to AutoCorres, but `asks` is the more standard name for people familiar with that monad. I think we should make one of them a translation. Maybe the following:

 - keep the definition of `ask` and `asks` (in `OptionMonad`)
 - make `ogets` a (full) abbreviation for `asks`
 - provide a lemma `ogets_def` in in `OptionMonadWP` that corresponds to the old definition

This should keep the name people know in AutoCorres, as well as the name for people how are familiar with the reader monad.

Reply:
ok, I’ll do that.